numba_dpex/tests/test_parfors.py

```diff
@@ -0,0 +1,887 @@
+                                 parameter is modified rather than a certain
+                                 value returned.  If this keyword argument
+                                 is supplied, it should be a list of
+                                 comparison functions such that the i'th
+                                 function in the list is used to compare the
+                                 i'th parameter of the njit and parallel=True
+                                 functions against the i'th parameter of the
+                                 standard Python function, asserting if they
+                                 differ.  The length of this list must be equal
+                                 to the number of parameters to the function.
+                                 The null comparator is available for use
+                                 when you do not desire to test if some
+                                 particular parameter is changed.
+            Remaining kwargs are passed to np.testing.assert_almost_equal
+        """
+        scheduler_type = kwargs.pop('scheduler_type', None)
+        check_fastmath = kwargs.pop('check_fastmath', None)
+        fastmath_pcres = kwargs.pop('fastmath_pcres', None)
+        check_scheduling = kwargs.pop('check_scheduling', True)
+        check_args_for_equality = kwargs.pop('check_arg_equality', None)
+
+        def copy_args(*args):
+            if not args:
+                return tuple()
+            new_args = []
+            for x in args:
+                if isinstance(x, np.ndarray):
+                    new_args.append(x.copy('k'))
+                elif isinstance(x, np.number):
+                    new_args.append(x.copy())
+                elif isinstance(x, numbers.Number):
+                    new_args.append(x)
+                elif x is None:
+                    new_args.append(x)
+                elif isinstance(x, tuple):
+                    new_args.append(copy.deepcopy(x))
+                elif isinstance(x, list):
+                    new_args.append(x[:])
+                else:
+                    raise ValueError('Unsupported argument type encountered')
+            return tuple(new_args)
+
+        # python result
+        print("*args:", *args)
+        py_args = copy_args(*args)
+        py_expected = pyfunc(*py_args)
+        print("py_expected:", py_expected)
+
+        # njit result
+        njit_args = copy_args(*args)
+        njit_output = cfunc.entry_point(*njit_args)
+        print("njit_output:", njit_output)
+
+        # parfor result
+        parfor_args = copy_args(*args)
+        parfor_output = cpfunc.entry_point(*parfor_args)
+        print("parfor_output:", parfor_output)
+
+        # numba_dpex_result
+        # TODO: what to do when we need to have a data parallel operations?
+        numba_dpex_args = copy_args(*args)
+        with dpctl.device_context(self.device):
```

Comment:
@diptorupd the test suite creates two functions, one with njit and another with parfor and they are tested against python output. Here what I am doing is running one of those functions through dpctl and looks like it's running. Now my question is:
1. How do I know if they are actually running on sycl device?
2. What if I need to do data parallel operations? i.e. `dpex.get_global_id(0)`

Reply:
>     1. How do I know if they are actually running on sycl device?

That is a question we need to solve properly. There is a fallback feature in dpex that executes code on CPU if offload failed. For the test case, set the `numba_dpex.config.
FALLBACK_ON_CPU` to `0`. Refer: https://github.com/IntelPython/numba-dpex/blob/192ef38e50a65974cfcd91757d802a54fc78bbb8/numba_dpex/config.py#L83

Doing so, will turn off fallback and we will get a failure message if offload did not work.

The other thing that we have to start doing is to figure out how to interface with onetrace (https://intelpython.github.io/dpctl/latest/docfiles/dpctl/dpctl.utils_functions_api.html#dpctl.utils.onetrace_enabled). The `dpctl.utils.onetrace_enabled` will output the memory address of the queue where the kernel is executing, we have to compare that with the memory address of the queue specified in `device_context`.


>     2. What if I need to do data parallel operations? i.e. `dpex.get_global_id(0)`

For the numpy and parfor cases, dpex auto-generates the kernel and any indexing that is needed. So, `dpex.get_global_id` calls will not be visible in the user written code for parfors.